flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/parse/CalciteParser.java

```diff
@@ -47,12 +49,36 @@ public CalciteParser(SqlParser.Config config) {
      * @param sql a sql string to parse
      * @return a parsed sql node
      * @throws SqlParserException if an exception is thrown when parsing the statement
+     * @throws SqlParserEOFException if the statement is incomplete
      */
     public SqlNode parse(String sql) {
         try {
             SqlParser parser = SqlParser.create(sql, config);
             return parser.parseStmt();
         } catch (SqlParseException e) {
+            if (e.getMessage().contains("Encountered \"<EOF>\"")) {
+                throw new SqlParserEOFException(e.getMessage(), e);
+            }
+            throw new SqlParserException("SQL parse failed. " + e.getMessage(), e);
+        }
+    }
+
+    /**
+     * Parses a SQL string into a {@link SqlNodeList}. The {@link SqlNodeList} is not yet validated.
+     *
+     * @param sql a sql string to parse
+     * @return a parsed sql node list
+     * @throws SqlParserException if an exception is thrown when parsing the statement
+     * @throws SqlParserEOFException if the statement is incomplete
+     */
+    public SqlNodeList parseSqlList(String sql) {
+        try {
+            SqlParser parser = SqlParser.create(sql, config);
+            return parser.parseStmtList();
+        } catch (SqlParseException e) {
+            if (e.getMessage().contains("Encountered \"<EOF>\"")) {
+                throw new SqlParserEOFException(e.getMessage(), e);
```

Comment:
personally I like explicit subclass better, it can make code easier to understand.

Reply:
That's personal preference and it's not a blocking issue for me. But i can see how we might want in future to catch other parsing errors (also see in context of https://cwiki.apache.org/confluence/display/FLINK/FLIP-189%3A+SQL+Client+Usability+Improvements) and one class per error type is overkill and it's going to explode. The point is that if what you need is just to match the specific error type variant, but the variants themselves don't have any other useful error data, then you don't subclasses.

Reply:
so can we keep the class for now and do some necessary refactor in the future when necessary?

Reply:
What are the implications if we do this now vs later? Also look at this point in particular: https://cwiki.apache.org/confluence/display/FLINK/FLIP-189%3A+SQL+Client+Usability+Improvements#FLIP189:SQLClientUsabilityImprovements-ImplementationDetails this talks about how in future we might want to extend `SqlMultiLineParser` to catch more parsing errors to leverage other jline features. Having a single sub exception for all of these is going to pollute our classpath.

Reply:
I think you may misunderstand the FLIP,  the FLIP want to add some quick check on the statement provided by jline, instead of using Flink Parser. I don't see the need to introduce Exception with ErrorType for now actually. If in the future we really need to introduce more exception, we can remove the EOFException, replace with the new one.